labelbox/data/serialization/ndjson/label.py

```diff
@@ -45,40 +81,51 @@ def from_common(cls,
             yield from cls._create_video_annotations(label)
 
     def _generate_annotations(
-        self,
-        grouped_annotations: Dict[str,
-                                  List[Union[NDObjectType, NDClassificationType,
-                                             NDConfusionMatrixMetric,
-                                             NDScalarMetric, NDSegments]]]
+        self, annotation_groups: Dict[str, _AnnotationGroupTuple]
     ) -> Generator[Label, None, None]:
-        for _, annotations in grouped_annotations.items():
-            annots = []
-            data_row = annotations[0].data_row
-            for annotation in annotations:
-                if isinstance(annotation, NDDicomSegments):
-                    annots.extend(
-                        NDDicomSegments.to_common(annotation, annotation.name,
-                                                  annotation.schema_id))
-                elif isinstance(annotation, NDSegments):
-                    annots.extend(
-                        NDSegments.to_common(annotation, annotation.name,
-                                             annotation.schema_id))
-                elif isinstance(annotation, NDDicomMasks):
-                    annots.append(NDDicomMasks.to_common(annotation))
-                elif isinstance(annotation, NDVideoMasks):
-                    annots.append(NDVideoMasks.to_common(annotation))
-                elif isinstance(annotation, NDObjectType.__args__):
-                    annots.append(NDObject.to_common(annotation))
-                elif isinstance(annotation, NDClassificationType.__args__):
-                    annots.extend(NDClassification.to_common(annotation))
-                elif isinstance(annotation,
+        for _, group in annotation_groups.items():
+            annotations = []
+            for uuid, ndjson in group.ndjson_objects.items():
+                if isinstance(ndjson, NDDicomSegments):
+                    annotations.extend(
+                        NDDicomSegments.to_common(ndjson, ndjson.name,
+                                                  ndjson.schema_id))
+                elif isinstance(ndjson, NDSegments):
+                    annotations.extend(
+                        NDSegments.to_common(ndjson, ndjson.name,
+                                             ndjson.schema_id))
+                elif isinstance(ndjson, NDDicomMasks):
+                    annotations.append(NDDicomMasks.to_common(ndjson))
+                elif isinstance(ndjson, NDVideoMasks):
+                    annotations.append(NDVideoMasks.to_common(ndjson))
+                elif isinstance(ndjson, NDObjectType.__args__):
+                    annotation = NDObject.to_common(ndjson)
+                    annotations.append(annotation)
+                    group.annotations[uuid] = annotation
```

Comment:
This looks like source/target can be any `NDObjectType`. Have you considered limiting it to only the object types that allow relationships, or would that introduce too much spaghetti?

Reply:
Also, if there are 100k object annotations for one data row, will we be storing two copies of them in memory here, in `annotations` from line 87, as well as `group.annotations`? Logically, it makes sense to have a separately stored mapping of `uuid` to `annotation`, so we can fetch them when we construct the relationship, but wondering if we've given considerations in terms of memory consumption

Reply:
It stores references to the same objects in two places, which is just an int.
Memory usage for both data structures (list and dict):
- 100K object annotations would require ~6MB
- 1M object annotations would require ~50MB
Let me know if this something I need to consider, to find another workaround for this.

Reply:
Isn't `NDObjectType` an Union of all the types that are currently supported by relationships? it contains `NDLine`, `NDPolygon`, `NDPoint`, `NDDocumentRectangle`, `NDRectangle`, `NDMask`, `NDConversationEntity`, `NDTextEntity`, `NDDocumentEntity`

Reply:
Ah gotcha. If they are both holding references to `annotation`, then I think we're fine

Reply:
Also, just verified and we do allow relationships on any object types. So this is good